Add Calico as a Networking Option

djosborne opened the pull request:
Looking for feedback on this PR to add Calico as an optional networking option for Openshift.

There are still a few missing pieces and some bugs, so I don't recommend anyone attempt to run this yet.

Isolation isn't working in this yet as the policy controller is not started. I will work towards adding it but just wanted to make sure my general approach was acceptable before doing so.

sdodson commented:
FYI @rajatchopra

rajatchopra commented:
As per a review meeting, the following three items need attention (in no particular order of severity/priority):

1. Make sure the use_calico option is mutually exclusive
2. Create a separate task for calico-master for the policy+controller service
3. Eventually move off to having a separate etcd instance (do not use the one that openshift creates)

@djosborne Thanks for the PR. Looking forward to getting this merged. Let me know if I missed anything.

djosborne commented:
@rajatchopra I'm trying to implement the 2nd suggestion (split into master / common role). But for some reason when I run with [the 5th commit in this PR](https://github.com/openshift/openshift-ansible/pull/3038/commits/8b98e4ded0060817294c8c6be731a7d9204826e3), the etcd certs become empty directories instead of ... certificates. I think it has to do with [the variables set in master/config.yml](https://github.com/openshift/openshift-ansible/pull/3038/commits/8b98e4ded0060817294c8c6be731a7d9204826e3#diff-385d17203ec610a738ee3ec5938ce287R177), but I'm not sure what I'm doing wrong.

I'm running this on a cluster where the host is both a master and node.

jcpowermac commented on roles/etcd_client_certificates/tasks/main.yml:
```diff
@@ -1,4 +1,24 @@
 ---
+- name: DEBUG etcd_cert_config_dir
```
@djosborne not sure if you caught this but you brought over some of my debug / pause actions.

djosborne commented:
@rajatchopra This PR should be ready for you to take a look. Here's some information on some  caveats:

### Sample Inventory.ini
Tested with Openshift origin v1.3
```
[OSEv3:children]
masters
nodes
etcd

[OSEv3:vars]
os_sdn_network_plugin_name=cni
openshift_use_openshift_sdn=false
openshift_use_calico=true

[masters]
master1 ansible_host=104.196.229.127

[nodes]
master1 openshift_node_labels="{'region': 'infra'}" 
node2 ansible_host=104.198.3.155

[etcd]
db1 ansible_host=104.196.253.167
```

### External etcd
This calico playbook only supports explicitly defined etcd hosts. I modeled it after the flannel playbook role which has the same requirement.

### Generating etcd Client Certs
The newly introduced Calico playbook role lists `openshift_etcd_client_certificates` as a dependency. But (with the help of @jcpowermac), I discovered that on master hosts, generation of the certs wasn't happening until _after_ calico's playbook role was executed. Because of this, Calico was volume mounting non-existent certs into it's docker container. Docker's default behavior is to then create empty directories on the host in this scenario.

To circumvent this, I modified the role to have the following behavior:

- Node-only hosts successfully generate etcd client certs which Calico will use.
- Master-only and master+node hosts configure Calico to use the existing etcd client certs on the host. Calico certs are still generated for these hosts afterwards, but are not used.

The correct fix here is probably to fix etcd client cert generation so that it happens before the Calico role. But I don't want to waste too much time on this since long term we will likely switch to Calico's "etcdless" backend which uses the kubernetes apiserver to store data, removing the dependency on etcd altogether.

### Etcd cert SAN field
[According to digicert](https://www.digicert.com/subject-alternative-name-compatibility.htm): 
>If a SSL Certificate has a Subject Alternative Name (SAN) field, then SSL clients are supposed to ignore the Common Name value and seek a match in the SAN list. This is why DigiCert always repeats the common name as the first SAN in our certificates.

I found that the SSL libraries Calico uses follows this same paradigm, and was skipping the SN of the etcd cert and jumping directly to the SAN, which only specified the IP of etcd. I configured Calico to connect to etcd using its hostname, so I added the hostname in the SAN field. 

### Other
- I haven't tested with multi-master yet. 
- I'm not completely satisfied with how the `calico` and `calico_master` role are split up / depend on each other. I think I'd like to introduce `calico_certs` and `calico_common` roles to better divy up, but figured since this "works" it'd be best to hold off for now.

zhwenh commented:
hello , I "git clone https://github.com/djosborne/openshift-ansible/tree/calico  " this version, and setup show this error:
TASK [openshift_ca : Create the master certificates if they do not already exist] ***
fatal: [10.1.130.15 -> 10.1.130.15]: FAILED! => {"changed": true, "cmd": ["/usr/local/bin/oc", "adm", "create-master-certs", "--hostnames=kubernetes.default,kubernetes.default.svc.cluster.local,kubernetes,openshift.default,openshift.default.svc,172.30.0.1,10.1.130.15,openshift.default.svc.cluster.local,kubernetes.default.svc,openshift", "--master=https://10.1.130.15:8443", "--public-master=https://10.1.130.15:8443", "--cert-dir=/etc/origin/master", "--overwrite=false"], "delta": "0:00:03.154188", "end": "2017-02-06 16:29:33.655000", "failed": true, "rc": 1, "start": "2017-02-06 16:29:30.500812", "stderr": "Command \"create-master-certs\" is deprecated, Use 'oc adm ca' instead.\nopen /etc/origin/master/master.etcd-client.crt: is a directory", "stdout": "Generated new key pair as /etc/origin/master/serviceaccounts.public.key and /etc/origin/master/serviceaccounts.private.key", "stdout_lines": ["Generated new key pair as /etc/origin/master/serviceaccounts.public.key and /etc/origin/master/serviceaccounts.private.key"], "warnings": []}

NO MORE HOSTS LEFT *************************************************************
        to retry, use: --limit @/root/ansible-calico/openshift-ansible-calico/playbooks/byo/config.retry

**my hosts :**
[OSEv3:children]
masters
nodes
etcd
# Set variables common for all OSEv3 hosts
[OSEv3:vars]
# SSH user, this user should allow ssh based auth without requiring a password
openshift_master_hostnames=10.1.130.15
ansible_ssh_user=root
openshift_clock_enabled=true
openshift_install_examples=true
#deployment_subtype=registry
deployment_type=origin

openshift_master_default_subdomain=svc.cluster.local
osm_cluster_network_cidr=172.88.0.0/16
openshift_release=v1.3.1
openshift_image_tag=v1.3.1
openshift_use_dnsmasq=true
openshift_pkg_version=-1.3.1
openshift_dns_ip=172.30.0.1
openshift_docker_options="--insecure-registry 172.30.0.0/16 --insecure-registry 172.88.0.0/16 --insecure-registry  10.1.0.0/16 --ipv6=false"
openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true', 'challenge': 'true', 'kind': 'HTPasswdPasswordIdentityProvider', 'filename': '/etc/origin/master/htpasswd'}]
openshift_additional_repos=[{'id': 'origin-release', 'name': 'origin-release', 'baseurl': 'http://cbs.centos.org/repos/paas7-openshift-origin-release/x86_64/os/','enabled': 1, 'gpgcheck': 0}]
os_sdn_network_plugin_name=cni
openshift_use_openshift_sdn=false
openshift_use_calico=true

# host group for masters
[masters]
10.1.130.15

[etcd]
10.1.130.15
# host group for nodes, includes region info
[nodes]
10.1.130.15 openshift_node_labels="{'nodetype': 'db'}" 
10.1.130.16 openshift_node_labels="{'nodetype': 'normal'}"
10.1.130.14 openshift_node_labels="{'nodetype': 'normal'}"

DanyC97 commented:
@zhwenh you mind if you edit your last comment and format the output into block code so it can be easily read?

DanyC97 commented:
will be nice to make merge this work, not sure what else is needed though

djosborne commented:
Rebasing onto latest head of `openshift-ansible` is causing connectivity issues when the cluster comes up with Calico. I'm investigating this week.

djosborne commented:
Hey all, I've fixed up this PR and believe it should work. Tested with Openshift 1.4 and it seems to be working.

One of the fixes required for 1.4 was ensuring the CNI loopback plugin was installed. I added this commit (https://github.com/openshift/openshift-ansible/pull/3038/commits/b183d971f693eb6c62a0da94c3a6862cb7c41ebe) but we should really come up with a more sensible way to resolve this correctly. I raised an issue at #3022 )

jcpowermac commented on roles/calico_master/tasks/main.yml:
```diff
@@ -0,0 +1,52 @@
+  changed_when: False
+
+
+- name: Set permissions for calico
```
@djosborne can you replace this `command` with the new `lib_openshift` module actions?
Example here: https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_hosted/tasks/registry/registry.yml#L60

https://github.com/openshift/openshift-ansible/tree/master/roles/lib_openshift/library

jcpowermac commented:
@djosborne once we get closer to approving this PR can this be squashed?

smarterclayton commented:
Is this using the master etcd?  If so the only access allowed will be from the master nodes - we don't let other clients talk directly to etcd - a separate etcd would be required.

djosborne commented:
@smarterclayton - This does access master etcd from the nodes. @rajatchopra originally OK'd this as a short term solution. Instead of doing the work to roll our own etcd, I plan to jump directly to using the Kuberentes Apiserver as a backend, removing Calico's dependency on etcd altogether. Planning this as follow-up work.

I'm doing some final testing but would love to see this get in as a first increment this week. Can someone provide some insight on why the travis-ci build is failing? I can't understand what is causing it.

I've live-demoed to @jcpowermac and addressed his PR feedback. What else is needed to get this in?

smarterclayton commented:
I don't know that we can recommend that as anything other than a trial or demo setup.  We don't allow that for other components, and so it's pretty important that this be:

a) impossible to enable without understanding the security implications
b) not be copied to other places in the code until it is removed within a reasonably short time window 
c) not listed as a recommended or supported configuration

@liggitt @sdodson Scott can you make sure that there is the appropriate security and controls around this so that it does not get kept as is or allowed to escape the very limited circumstances here?  Also a timeframe for this to move to not having direct access.

jcpowermac commented:
@djosborne looks like the reason travis is failing is:
`./roles/openshift_facts/library/openshift_facts.py:469:1: E302 expected 2 blank lines, found 1`

djosborne commented:
@smarterclayton 

>I don't know that we can recommend that as anything other than a trial or demo setup.

That is fine. The intent here is to get a base integration working. This work has been going on since before 1.4 was released and has taken unexpetedly long to get working. I'd really like to see this merged as a base working case with the intent of immediately iterating on and improving it.

jcpowermac commented:
@djosborne maybe remove WIP so the maintainers know this is ready.
@sdodson or @abutcher can you please review.

rajatchopra commented:
@djosborne The PR looks fine to me. Two things:

1. We are planning to restructure the ansible repo for plugins. So the way this is done is likely going to be rejigged. This PR should nevertheless get merged as the re-architecture starts, but expect to work on the new model soon.
2. The etcd being re-used is a non-flyer for production as you have understood. Could you add a README to the calico folder indicating that its not intended for production. Also link to some docs summarizing calico plugin so that any community user has at least some clues on what/how/why things work with the calico plugin.

djosborne commented:
@rajatchopra I've added the warning, and opened an issue at Calico Docs to add OpenShift documentation resources. Will move that along ASAP

sdodson commented on roles/calico/tasks/main.yml:
```diff
@@ -0,0 +1,74 @@
+  with_items:
+  - "{{ calico_etcd_ca_cert_file }}"
+  - "{{ calico_etcd_cert_file}}"
+  - "{{ calico_etcd_key_file }}"
```
You're not actually creating these, you're requiring that they exist prior to running ansible? Perhaps document how you'd expect to create them in README.md?

sdodson commented on roles/calico_master/tasks/main.yml:
```diff
@@ -0,0 +1,41 @@
+  with_items:
+  - "{{ calico_etcd_ca_cert_file }}"
+  - "{{ calico_etcd_cert_file}}"
+  - "{{ calico_etcd_key_file }}"
```
Likewise, document how to create these?

sdodson commented on roles/calico/templates/calico.service.j2:
```diff
@@ -0,0 +1,29 @@
+ -v /var/log/calico:/var/log/calico \
+ -v /lib/modules:/lib/modules \
+ -v /var/run/calico:/var/run/calico \
+ calico/node:v1.1.0
```
Is this on quay.io like calico-policy-controller image? If so you'll need to specify it as it's not included in our default docker config.

sdodson commented:
aos-ci-test

sdodson commented:
[test]

sdodson commented:
At some point in the future origin will support arm64 and ppc64le architectures. Which architectures is calico built for?

sdodson commented:
[merge]

zhwenh commented:
calico network failed to "SetupNetwork" 
 
origin v1.5.0-rc.0 Version + Calico + cni

Steps To Reproduce

router-3-deploy_default pod run show error message

Current Result

Expected Result

Additional Information

Error syncing pod, skipping: failed to "SetupNetwork" for "router-3-deploy_default" with SetupNetworkError: "Failed to setup network for pod "router-3-deploy_default(799d976f-1502-11e7-b3c1-fa163e10ef43)" using network plugins "cni": failed to send CNI request: Post http://dummy/: dial unix /var/run/openshift-sdn/cni-server.sock: connect: no such file or directory; Skipping pod"	39 times in the last 0 seconds

sdodson commented:
@zhwenh Can you please open an issue for that and mention @djosborne ? Your complete inventory would be helpful too.
